### PR TITLE
feat: add "/http" and "/graphql" export paths

### LINKED ISSUE #2038: Add "http" and "graphql" exports
### Scope

Adds a new behavior

### Compatibility

- [ ] This is a breaking change

### Feature description

- Originally discussed in https://github.com/mswjs/msw/discussions/2037

### PR DESCRIPTION
**Pull request by thepassle**
… file

- Discussed in #2037 (#1984)
- Fixes #2038

### CODE REVIEW

File: package.json
Comment (kettanaito) [inline package.json]: We seem to have inconsistent `default` condition across exports:

- `/http`, `/graphql` and `/node` list the ESM build as the default;
- `.`, `/browser` and `/native` list the CJS build as the default. 

Is this something we should unify as a part of this change as well?

Comment (thepassle) [reply package.json]: we can change the `default` for `./http` and `./graphql`, but changing the defaults for other keys may be a breaking change

Comment (kettanaito) [reply package.json]: I was mainly suggesting the former. As we are in this weird CJS/ESM situation, the default must remain CJS. Would you please make the change?

Comment (thepassle) [reply package.json]: done

**Review by kettanaito — DISMISSED** [review]
This looks great! Let's have this released in the next minor version. **Just don't merge this yet**, I'd like to release a few bug fixes before the next version bump. Thanks. 

@thepassle, how do you recommend us mention these exports in the docs? I still think that importing from the root should be the default. I wonder where these export paths would be proper to mention.

**Comment by pleunv** [comment]
The annoying thing is that autocompletion will most likely favor the shortest path, and I don't think there's a way to tell it otherwise. What I generally end up doing is set up eslint's [no-restricted-imports](https://eslint.org/docs/latest/rules/no-restricted-imports) rule with a custom message to reroute users to the right import. Not the most user-friendly approach, though.

Something along the lines of:

```json
"rules": {
  "no-restricted-imports": [
    "error",
    {
      "paths": [
        {
          "name": "msw",
          "importNames": ["graphql"],
          "message": "Please import 'graphql' from 'msw/graphql' instead."
        },
      ]
    }
  ]
}
```

**Comment by kettanaito** [comment]
@pleunv, autocomplete right now is pretty broken for me anyways. VS Code always tries to import things where their _type definitions are_, not what the export says in package.json. For example, when I autoimport `http` I get this:

```ts
import { http } from 'msw/lib/http'
```

Which is incorrect; you shouldn't be importing things from `/lib/`. This looks like a problem the end developer has to configure themselves.

**Comment by pleunv** [comment]
Of all the TypeScript issues I've ever run into, fortunately that's not one of them 😄. That would honestly drive me mad. Did you override any of the "Import Module" settings?

**Comment by kettanaito** [comment]
@pleunv, no, that's the default behavior.

**Comment by thepassle** [comment]
> This looks great! Let's have this released in the next minor version. Just don't merge this yet, I'd like to release a few bug fixes before the next version bump. Thanks.

Okidoki, any idea on when the next minor will be released?

> I still think that importing from the root should be the default. I wonder where these export paths would be proper to mention.

Not really sure why you wouldnt change this:
image

to:
image

I hardly think this impacts DX for the largest usecase, even if you use both graphql and http, but I'm also not really interested in discussing that any longer; I think the perf difference outweighs DX, but it seems I haven't been able to convince you of that, so with peace and love, I won't go into that any further and leave it at this :)

> @thepassle, how do you recommend us mention these exports in the docs? 

Ill leave that up to your discretion

**Comment by kettanaito** [comment]
> Okidoki, any idea on when the next minor will be released?

I'm releasing next 2 bug fixes related to React Native, and then it's the minor release time!

> Not really sure why you wouldnt change this: ... to:

You are omitting the experience implications here. All the tutorials, videos, and articles feature `from 'msw'`. If the person goes to the docs and, suddenly, it's `from 'msw/http'`, a red flag goes in their head that someone is wrong. Either the materials they read are wrong or the docs. 

That's why performance improvements like these deserve a separate page, I think. Before (and if) we drop the root-level export altogether, I won't recommend this by default for one simple reason: nobody has reported performance issues due to barrel imports in the past, so nobody had performance issues due to barrel imports, it's as simple as that. 

Those who need this fix should be able to find it. 

> but it seems I haven't been able to convince you of that

You've convinced me. From my side, I'm trying to show you the area of effect this change has on the end developer. I will not discard educational experience and simplicity of use. The `msw/http` and `msw/graphql` imports will be _opt-in_ for those who experience issues. Those who don't can keep using `msw` root-level export as they did in the past.

**Comment by kettanaito** [comment]
I have one piece of feedback. The format we have now is `msw/<environment>`: `msw/node`, `msw/browser`, `msw/native`. I feel that `msw/http` breaks this rule. 

What do you think about `msw/core/http`/`msw/core/graphql`?

**Comment by kettanaito** [comment]
## Released: v2.2.0 🎉

This has been released in v2.2.0!

- 📄 [**Release notes**](https://github.com/mswjs/msw/releases/tag/v2.2.0)
- 📦 [npm package](https://www.npmjs.com/package/msw/v/2.2.0)

Make sure to always update to the latest version (`npm i msw@latest`) to get the newest features and bug fixes.

---

_Predictable release automation by [@ossjs/release](https://github.com/ossjs/release)_.

**Pull request merged**
